File: hdar/cpp/src/remote_ssh.cpp
#include "hdar/remote_ssh.hpp"
#include <cstdio>
#include <array>
#include <chrono>
#include <filesystem>
#include <sstream>

namespace fs = std::filesystem;

namespace hdar {

RemoteSshProvider::RemoteSshProvider(const std::string& host,
                                      const std::string& ssh_user,
                                      const std::string& remote_base,
                                      int ssh_port)
    : host_(host)
    , ssh_user_(ssh_user)
    , remote_base_(remote_base)
    , ssh_port_(ssh_port) {}

std::string RemoteSshProvider::remote_dir(const std::string& runtime_id) const {
    return remote_base_ + "/" + runtime_id;
}

std::string RemoteSshProvider::ssh_cmd(const std::string& remote_command) const {
    std::string cmd = "ssh";
    if (ssh_port_ != 22) cmd += " -p " + std::to_string(ssh_port_);
    if (!ssh_user_.empty()) cmd += " " + ssh_user_ + "@";
    else cmd += " ";
    cmd += host_ + " '" + remote_command + "'";
    return cmd;
}

std::string RemoteSshProvider::scp_cmd(const std::string& local, const std::string& remote) const {
    std::string cmd = "scp";
    if (ssh_port_ != 22) cmd += " -P " + std::to_string(ssh_port_);
    if (!ssh_user_.empty()) cmd += " " + local + " " + ssh_user_ + "@" + host_ + ":" + remote;
    else cmd += " " + local + " " + host_ + ":" + remote;
    return cmd;
}

std::string RemoteSshProvider::exec_cli(const std::string& cmd, int* exit_code) {
    std::string result;
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) {
        if (exit_code) *exit_code = -1;
        return "";
    }
    std::array<char, 4096> buf;
    while (fgets(buf.data(), buf.size(), pipe) != nullptr)
        result += buf.data();
    int status = pclose(pipe);
    if (exit_code) *exit_code = WEXITSTATUS(status);
    return result;
}

RuntimeRecord RemoteSshProvider::materialize(
    const std::string& runtime_id,
    const std::string& workspace_path,
    const std::string& image,
    const std::string& cpu_limit,
    const std::string& memory_limit,
    const std::string& network_policy) {

    RuntimeRecord rec;
    rec.provider = "remote-ssh";
    rec.runtime_id = runtime_id;
    rec.image_digest = image;
    rec.cpu_limit = cpu_limit;
    rec.memory_limit = memory_limit;
    rec.workspace_mount = workspace_path;
    rec.network_policy = network_policy;
    rec.start_timestamp = epoch_seconds();

    std::string rdir = remote_dir(runtime_id);

    int rc = -1;
    exec_cli(ssh_cmd("mkdir -p " + rdir + "/workspace"), &rc);

    if (rc != 0) {
        rec.exists = false;
        runtimes_[runtime_id] = rec;
        return rec;
    }

    if (fs::exists(workspace_path) && fs::is_directory(workspace_path)) {
        std::string tar_cmd = "tar cf - -C " + workspace_path + " . | " +
                              ssh_cmd("tar xf - -C " + rdir + "/workspace");
        exec_cli(tar_cmd, &rc);
    }

    rec.exists = true;
    rec.vm_identity = host_ + ":" + rdir;
    runtimes_[runtime_id] = rec;
    return rec;
}

ExecutionResult RemoteSshProvider::execute(
    const std::string& runtime_id,
    const std::string& operation_type,
    const std::string& command,
    int timeout) {

    ExecutionResult result;
    result.operation_type = operation_type;
    result.command = command;

    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end() || !it->second.exists) {
        result.stderr_text = "runtime not found or not running";
        return result;
    }

    std::string rdir = remote_dir(runtime_id);
    std::string cmd = ssh_cmd("cd " + rdir + "/workspace && " + command);

    auto start = std::chrono::steady_clock::now();
    int rc = -1;
    result.stdout_text = exec_cli(cmd, &rc);
    result.exit_code = rc;
    result.success = (rc == 0);

    auto end = std::chrono::steady_clock::now();
    result.duration_ms = std::chrono::duration<double, std::milli>(end - start).count();

    return result;
}

RuntimeRecord RemoteSshProvider::stop(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    it->second.stop_timestamp = epoch_seconds();
    return it->second;
}

RuntimeRecord RemoteSshProvider::destroy(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    std::string rdir = remote_dir(runtime_id);
    int rc = -1;
    std::string del_output = exec_cli(ssh_cmd("rm -rf " + rdir), &rc);

    it->second.delete_timestamp = epoch_seconds();
    it->second.exists = false;

    JsonValue inspection = JsonValue::object();
    inspection["exists"] = JsonValue::boolean(false);
    inspection["provider"] = JsonValue::string("remote-ssh");
    inspection["runtime_id"] = JsonValue::string(runtime_id);
    inspection["remote_dir"] = JsonValue::string(rdir);
    inspection["delete_exit_code"] = JsonValue::integer(rc);
    it->second.post_delete_inspection = inspection;

    return it->second;
}

JsonValue RemoteSshProvider::inspect(const std::string& runtime_id) {
    JsonValue v = JsonValue::object();
    v["provider"] = JsonValue::string("remote-ssh");
    v["runtime_id"] = JsonValue::string(runtime_id);

    std::string rdir = remote_dir(runtime_id);
    int rc = -1;
    exec_cli(ssh_cmd("test -d " + rdir + " && echo EXISTS"), &rc);

    v["exists"] = JsonValue::boolean(rc == 0);
    v["remote_dir"] = JsonValue::string(rdir);
    v["host"] = JsonValue::string(host_);

    auto it = runtimes_.find(runtime_id);
    if (it != runtimes_.end()) {
        v["start_timestamp"] = JsonValue::number(it->second.start_timestamp);
        if (it->second.stop_timestamp)
            v["stop_timestamp"] = JsonValue::number(*it->second.stop_timestamp);
        if (it->second.delete_timestamp)
            v["delete_timestamp"] = JsonValue::number(*it->second.delete_timestamp);
    }

    return v;
}

std::vector<std::string> RemoteSshProvider::list_runtimes() {
    std::vector<std::string> result;
    int rc = -1;
    std::string output = exec_cli(
        ssh_cmd("ls -1 " + remote_base_ + " 2>/dev/null"), &rc);

    if (rc == 0) {
        std::istringstream iss(output);
        std::string line;
        while (std::getline(iss, line)) {
            while (!line.empty() && (line.back() == '\r' || line.back() == '\n'))
                line.pop_back();
            if (!line.empty())
                result.push_back(line);
        }
    }
    return result;
}

} // namespace hdar

File: hdar/hdar_core/providers/remote_ssh.py
"""Remote SSH provider — for cross-host execution.

Transfers capsule to a remote host via SSH, restores it there,
executes operations, and collects signed execution receipts.
The remote host does NOT receive the owner's private key.
"""

from __future__ import annotations

import os
import subprocess
import time
from pathlib import Path
from typing import Optional

from .base import ProviderBase, RuntimeRecord, ExecutionResult


class RemoteSSHProvider(ProviderBase):
    """Execute on a remote host via SSH.

    The remote host receives:
    - capsule files
    - owner public key
    - expected agent_id and epoch
    - fenced lease and token
    - attenuated capability set

    It does NOT receive the owner private key.
    """

    def __init__(self, host: str, user: str = "capsule-agent",
                 remote_workspace_root: str = "/tmp/capsule-runtime",
                 ssh_key: Optional[str] = None):
        self.host = host
        self.user = user
        self.remote_root = remote_workspace_root
        self.ssh_key = ssh_key
        self._runtimes: dict[str, RuntimeRecord] = {}

    @property
    def name(self) -> str:
        return "remote-ssh"

    def _ssh_base(self) -> list:
        cmd = ["ssh"]
        if self.ssh_key:
            cmd.extend(["-i", self.ssh_key])
        cmd.append(f"{self.user}@{self.host}")
        return cmd

    def _scp_to(self, local: str, remote: str) -> subprocess.CompletedProcess:
        cmd = ["scp"]
        if self.ssh_key:
            cmd.extend(["-i", self.ssh_key])
        cmd.extend([local, f"{self.user}@{self.host}:{remote}"])
        return subprocess.run(cmd, capture_output=True, text=True, timeout=120)

    def materialize(
        self,
        runtime_id: str,
        workspace_path: str,
        image: str = "",
        cpu_limit: str = "2",
        memory_limit: str = "2g",
        network_policy: str = "none",
    ) -> RuntimeRecord:
        remote_dir = f"{self.remote_root}/{runtime_id}"

        # Create remote directory
        subprocess.run(
            self._ssh_base() + [f"mkdir -p {remote_dir}/workspace"],
            capture_output=True, text=True, timeout=30,
        )

        # Transfer workspace
        import tarfile
        import io
        import tempfile

        with tempfile.NamedTemporaryFile(suffix=".tar.gz", delete=False) as tmp:
            with tarfile.open(tmp.name, "w:gz") as tar:
                tar.add(workspace_path, arcname=".")
            self._scp_to(tmp.name, f"{remote_dir}/workspace.tar.gz")
            os.unlink(tmp.name)

        # Extract on remote
        subprocess.run(
            self._ssh_base() + [
                f"cd {remote_dir} && mkdir -p workspace && "
                f"tar xzf workspace.tar.gz -C workspace && rm workspace.tar.gz"
            ],
            capture_output=True, text=True, timeout=60,
        )

        record = RuntimeRecord(
            provider=self.name,
            runtime_id=runtime_id,
            image_digest=image or "remote:linux",
            vm_identity=f"{self.user}@{self.host}",
            cpu_limit=cpu_limit,
            memory_limit=memory_limit,
            workspace_mount=f"{remote_dir}/workspace",
            network_policy=network_policy,
            start_timestamp=time.time(),
            exists=True,
        )
        self._runtimes[runtime_id] = record
        return record

    def execute(
        self,
        runtime_id: str,
        operation_type: str,
        command: str,
        timeout: int = 60,
    ) -> ExecutionResult:
        if runtime_id not in self._runtimes:
            return ExecutionResult(
                operation_type=operation_type,
                command=command,
                exit_code=-1,
                stderr="runtime not found",
                success=False,
            )

        remote_dir = f"{self.remote_root}/{runtime_id}"
        start = time.time()
        result = subprocess.run(
            self._ssh_base() + [f"cd {remote_dir}/workspace && {command}"],
            capture_output=True, text=True, timeout=timeout,
        )
        duration = (time.time() - start) * 1000

        return ExecutionResult(
            operation_type=operation_type,
            command=command,
            exit_code=result.returncode,
            stdout=result.stdout,
            stderr=result.stderr,
            duration_ms=duration,
            success=result.returncode == 0,
        )

    def stop(self, runtime_id: str) -> RuntimeRecord:
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        record = self._runtimes[runtime_id]
        record.stop_timestamp = time.time()
        return record

    def destroy(self, runtime_id: str) -> RuntimeRecord:
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        record = self._runtimes[runtime_id]
        remote_dir = f"{self.remote_root}/{runtime_id}"

        subprocess.run(
            self._ssh_base() + [f"rm -rf {remote_dir}"],
            capture_output=True, text=True, timeout=30,
        )
        record.delete_timestamp = time.time()
        record.exists = False
        record.post_delete_inspection = self.inspect(runtime_id)
        del self._runtimes[runtime_id]
        return record

    def inspect(self, runtime_id: str) -> dict:
        remote_dir = f"{self.remote_root}/{runtime_id}"
        result = subprocess.run(
            self._ssh_base() + [f"test -d {remote_dir} && echo EXISTS || echo GONE"],
            capture_output=True, text=True, timeout=10,
        )
        exists = "EXISTS" in result.stdout
        return {
            "exists": exists,
            "provider": self.name,
            "runtime_id": runtime_id,
            "remote_path": remote_dir,
        }

    def list_runtimes(self) -> list:
        result = subprocess.run(
            self._ssh_base() + [f"ls {self.remote_root} 2>/dev/null || true"],
            capture_output=True, text=True, timeout=10,
        )
        return [d.strip() for d in result.stdout.splitlines() if d.strip()]
